scripts/automator/main.sh

```diff
@@ -0,0 +1,120 @@
+#!/bin/bash
+
+# Copyright (c) 2019 Kevin R Croft <krcroft@gmail.com>
+# SPDX-License-Identifier: GPL-2.0-or-later
+
+# Automator is a generic automation tool that helps separate data
+# from logical code.  It uses one or more "variables" files placed
+# in a single sub-directory, that are sourced in logical succession,
+# such as variables defined in earler files are either added-to or
+# overridden in subsequent files.
+#
+# For a thorough example, see scripts/build.sh and its data files
+# in scripts/automator/build.
+#
+# TODO: This script manages variables using eval meta-programming.
+#       Although the syntax is quite readable under bash 4.x,
+#       we have limitted our syntax to the uglier and heavier bash 3.x
+#       because Apple is still only shipping bash 3.x, even on their
+#       latest operating system (Nov-2019).
+#
+#       Future work might involve switching this to python and using YAML
+#       syntax in the variables files. Python's dictionaries can be
+#       merged or have their values overridden, and key names can be readily
+#       manipulated.
```

Comment:
Apple will never update bash, they're more likely to replace it with a different shell altogether.

I actually like bash scripting and various hacky ways you can abuse basically type-less system, but… if this scripts starts it's life with assumption, that it should probably be replaced by python code, then it probably be replaced by a python code right away.

And various techniques used in this script will make it barely untouchable for future contributors… so they will probably write their own script to wrap this one, and it will go on, and on ;) (been there, done that, I don't want a repeat)

Also, I would caution you before adding more complexity and simply use Python dictionaries instead of configuration files - with yaml configuration files I would question why we need them in the first place - we have already CI implemented as yaml files.

Reply:
I don't have the stamina to rewrite this engine in python just to get this PR through; it's not very trivial and would need another week or more to get it working.  That said, I do want to replace it with python already, and would open an issue to do it.  Having this as a working baseline is also important. 

regarding yaml in github CI vs yaml in python has less to do with how we store data and much more to do with functionality offered by the interpreted languages in each.  

In github CI, our programming logic is massively limited to cheesy 'matrix' loops and adding simple if conditions.. plus our 'code' itself is formatted into yaml.  We can't even store, use, and update variables.

Where as python, bash, etc.. gives us full flexibility.  So yaml here is merely an externalized data structure.  dicts are fine too..I just want to stay away from having the data infect the code with it's own logic.  Often when to force the data into some agnowtic structure, you're forced to create a generic solution process the data.

Reply:
> Apple will never update bash, they're more likely to replace it with a different shell altogether.

I can't see how any any self-respecting programmer would want to support and be part of the "Apple ecosystem". It sounds like they make developers walk on an apple-licensed tight-rope, with knives and an acid bath beneath if you dare stray. Get me out of there :fearful:!

Reply:
> I don't have the stamina to rewrite this engine in python to get this PR through; it's not very trivial and would need another week or more to get it working.

I understand… How about this: I will be more lenient with rest of review and will gladly help with a rewrite, but we would need to agree to *no new features* landing in these scripts before rewrite to Python is complete.

> Where as python, bash, etc.. gives us full flexibility. So yaml here is merely an externalized data structure.

The reason why I suggested to use python dicts instead of parsing files in some usual format is because in last ~year I faced the same decision twice in other projects, and after much deliberation, both times I decided to go with dicts, this code is open source so I can even link it, it was [here](https://github.com/ValveSoftware/Proton/pull/1423/files) and [here](https://github.com/dreamer/boxtron/blob/master/tweaks.py) (and one more time in closed project, where I went with saving configuration files in *.json, and then regretted it and once when I went with saving as *.ini files using built-in Python module and faced a number of unexpected bugs in result). The reasons:

- No need for additional parsing infrastructure, works on any machine without additional dependencies (parser is Python parser, there's no need to using pip to install additional modules (which would slow down CI btw)
- There's no need for including additional linters or checkers in CI to point out user errors in these files: pylint does that already, reporting exact lines with errors, editors handle syntax highlight correctly, there's no need to remember what types and what quirks of syntax of specific format are important
- There's no need to write and test code, that looks for files to parse and handles errors, etc - just import a python module and use a global const inside
- Functionality is basically the same